0101-nix formatting.md

```diff
@@ -0,0 +1,672 @@
+It is bound to the style guide and rules specified in this RFC.
+
+Should new syntax features be introduced into Nix, the formatter team should be consulted prior to their introduction.
+
+The team initially consists of:
+- @infinisil
+- @tomberek
+- @piegames
+- @0x4A6F
+- @kamadorueda (if available, TODO: @tomberek sent a message)
+
+The team has the authority to remove and add members as needed.
+
+## Examples and Interactions
+
+[examples-and-interactions]: #examples-and-interactions
+
+TODO sort that in somewhere else
+```nix
+[
+  "--some-flag" "some-value"
+]
+```
+
+is formatted as
+
+```
+[
+  "--some-flag"
+  "some-value"
+]
+```
+
+workaround:
+
+```
+lib.cli.toGNUCommandLine {} {
+  some-flag = "some-value";
+}
+```
+
+### Documentation
+
+Update [Section 20.1 (Coding conventiones → Syntax)](https://nixos.org/manual/nixpkgs/stable/#sec-syntax) of the Nixpkgs manual. All formatting-specific guidance is removed and replaced with instructions on how to automatically format Nixpkgs instead.
+
+### Automatically generated files
+
+TODO discuss that https://github.com/NixOS/rfcs/pull/101#discussion_r690272696
+- exclude option in tooling
+- 2nix apply fomatting style
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+- Having a commit that changes the formatting, can make git blame harder to use. It will need `--ignore-rev` flag.
+- Every formatter will have bugs.
+
+## Alternatives
+
+[alternatives]: #alternatives
+
+- Keep the status quo of not having an official formatter. The danger is that this creates discord within the nix community. On top of fragmented the community it can generate lengthy discussions (which do not advance the eco-system).
+
+### Incremental migration
+
+A more gradual approach, where only changed files get formatted for some period of time to make a more graceful transition has been rejected for practical concerns. Among other things, such an approach would not improve the amount of merge conflicts, and increase the workload for contributors during that time significantly.
+
+## Unresolved questions
+
+[unresolved]: #unresolved-questions
+
+- Are there situation where automated formatting is worse than manual formatting? Do we want to have exceptions to automated formatting?
+
+## Future work
+
+[future]: #future-work
+
+- TODO General style guidelines beyond AST reformatting
+
+## Appendix A: Formatting style guidelines
+
+**Nomenclature:**
+Brackets: `[]`
+Braces: `{}`
+Parentheses: `()`
+
+*These are not strict rules but guidelines to help create more consistent formatting rules.
+Nevertheless, deviations should be documented and explained.*
+
+- When deciding between two *equally good* options, currently prevalent formatting style in `nixpkgs` should be followed.
+- On the top level (indentation zero), special rules may apply.
+  - TODO: Is this still relevant?
+- Two spaces are used for each indentation level, there is no vertical alignment
```

Comment:
```suggestion
- One tab is used for each indentation level, there is no vertical alignment
```

Tabs have real world benefits for making the code more accessible for the visually impaired. 

I already outlined this a bit [here](https://github.com/NixOS/rfcs/pull/101#issuecomment-1707814452) .
I believe this is a point we should at least discuss.

Reply:
I am personally a huge fan of tabs for indentation too, and when I started Nix I used tabs for my configuration at first, and even kept it that way for years until [only nine months ago](https://git.darmstadt.ccc.de/piegames/home-config/-/commit/27ace65e09aa75d29dc2fe3c62678ff95ead5916). The big problem with tabs is that they are not supported in Nix strings in the same way spaces are. Using tabs for indentation means you must deal with that fact in some way.

In my home configuration, I created a small [`stripTabs`](https://git.darmstadt.ccc.de/piegames/home-config/-/commit/bfd7905d08b89f96b7a8cfcbf30eec8cfb83a132) helper function which emulates the "smart" behavior of Nix strings. This has the disadvantage that one must wrap every single multi line string in a function call. This is easy to forget (for example when a single line string becomes multi line), easy to miss and can cause really confusing behavior. (Most of the time things will work just fine and just look a bit weird should you happen to look at the generated output, until they don't.)

> as whitespace on strings have semantics that cannot be understood by a nix formatter

This is wrong. *Because* of the semantics of strings in Nix, indentation rules also apply *within* a string. The formatter is allowed to change characters within the string as long as it still evaluates to the same value. This also applies to interpolation and escaping by the way. Those strings only work with space indentation, thus unfortunately forcing that onto the rest of the language.

> literally zero cost

Please show me a zero-cost solution for how to use Tabs with Nix, preferably in a non-trivial production code base. I've shown you mine, and it was not zero cost at all. (I do know of some other ways to hack around this, but they all have their own significant downsides.)

-------

I don't see a way around using spaces for indentation in the current state of the Nix language, but I'd love to see the language fixed one day and then I'd be happy to move everything over to tabs. Unfortunately, this will likely require backwards-incompatible changes and thus depend on something like [RFC 137](https://github.com/NixOS/rfcs/pull/137)

Reply:
See also https://github.com/NixOS/nix/issues/7834

Reply:
Tabs over spaces forever tbh, way more easier to do and just easier to deal with imo.

Reply:
> I'd love to see the language fixed one day and then I'd be happy to move everything over to tabs. 

I don't think we can move over to tabs after this is finished, we would have to literally change every line of code in Nixpkgs again.

I'd still be in favor of tabs, but:
To use tabs as indentation, we have to either wait with this RFC or use tabs for everything outside multiline strings until we have support for tabs in multiline strings.
Since we don't actually know if tab support will ever happen, betting on being able to make it more consistent later is not preferable in my opinion.
Or we just accept the inconsistency (spaces in multiline strings and tabs everywhere else) and just hope we can make it more consistent without betting on it.

Reply:
> Please show me a zero-cost solution for how to use Tabs with Nix

Just to clarify, my argument was not regarding the semantics that tabs vs strings have regarding the nix codebase. My point is rather that tabs have little to no downside to the average human code reader / writer. I agree I could have phrased that better.

Reply:
Much of this talk seems to be off topic for this RFC, no?

Reply:
Preferences aside, I don't think using tabs in nixpkgs is feasible, since there are only 2 options I see

- Use spaces as a workaround whenever idented strings are used - This makes the code only look right at one tab width, which defeats the purpose of using tabs in the first place
- Fix Nix and raise the minver - Keep in mind the nixpkgs's minver is still at 2.3, adopting tabs would require the fix to be implemented in Nix, and we would need to raise the minver to a version that supports flakes, which would open a new can of worms that will probably take a very long time to reach an agreement.

Reply:
> Preferences aside

I think that's part of the problem right there. This discussion is not about preference. For someone in a wheelchair, using a ramp instead of the stairs is not a matter of preference.
This argument is about whether or not we want the nixpkgs code to be more accessible and this code formatting standard should reflect that decision.

Reply:
My point is not that we should not switch to tabs, but that we should not wait for the fix in nix and a debate about minver to make the decision on this RFC. I think it's perfectly fine to list this in unresolved questions and move on with this RFC since this is not a regression (we are already enforcing 2-space indentation).

Reply:
> I don't think we can move over to tabs after this is finished, we would have to literally change every line of code in Nixpkgs again.

Well yes. We are already going to touch every single code line now with this RFC, and we could always do another RFC for moving towards tabs and do it again. I'd be willing to do that part, but I won't be the one to fix the language, which is a prerequisite.

> To use tabs as indentation, we have to either wait with this RFC or use tabs for everything outside multiline strings until we have support for tabs in multiline strings.

This is a common suggestion, and additionally to the aforementioned problems it also does not interact well with nested string interpolations. Nix code can easily have three levels of interpolation nesting. Good luck properly dealing with that as a sighted user, not even speaking of how to deal with mixed tabs and spaces using a screen reader (it's possible, but a PITA).

Also, delaying this RFC even further for this reason is a non-starter IMO, especially since we are talking about a delay of probably at least several years. Formatting Nixpkgs is long overdue, don't let perfect be the enemy of the good here.

> My point is rather that tabs have little to no downside to the average human code reader / writer.

Generally speaking, yes. But not in Nix. Again, please show me a Nix code base that uses tabs for indentation without too much insanity. I am open to the idea, just very very skeptical because of my prior attempts.

Reply:
> Formatting Nixpkgs is long overdue, don't let perfect be the enemy of the good here.

Totally agree on that. It's been one of my major pain points when getting code review for pull requests on nixpkgs.
We could throw `alejandra` onto the code base tomorrow and be done with good enough.
I seems rather this RFC has gone the down the perfect road long before I chimed in here so we might as well finish in style.

Reply:
> We could throw alejandra onto the code base tomorrow and be done with good enough.

IMO, the current output from alejandra is "good enough". Personally, I define "good enough" as "better than the current state of Nixpkgs of average, and where better manual formatting could be achieved it wouldn't be worth the time and effort". For me, the nixfmt branch has reached that goal some time ago, and since then further improvements have brought it into a state where it is hopefully good enough for more critical people. But this is off-topic for this thread.

Reply:
> What I meant is, that the community is not likely to accept another big Nixpkgs refactor just for the sake of changing indentation to tabs. But this is probably the best option.

As a random community contributor (who is weakly pro-spaces—I won't prosecute the argument here, but it's more than just ‘I like it better’), I would be much more indulgent toward a future big refactor switching to tab indentation than I would toward just about any future large-scale tweak to some other formatting decision. There are technical pros and cons to the tabs-versus-spaces conversation; switching initial whitespace from one to the other is easy to automate even without building and running the latest iteration of the blessed formatter; and unlike most formatting decisions, it's entirely orthogonal to any other change that is likely to be in flight.

Reply:
To be honest I feel like the team responsible for formatting should just have carte blanche to alter formatting rules and format the tree accordingly, within reason (e.g. please batch it up so you're not doing it more than once per every couple releases at least!); the process of this RFC has demonstrated pretty well that the RFC format is not conducive to bikeshedding formatting decisions. With `.git-blame-ignore-revs`, the main issue is merge conflicts in PRs, which is admittedly not nothing, but is also a fact of life when contributing to Nixpkgs in general. So I would personally be content to see a change like this done without excessive amounts of red tape.

Reply:
In my opinion, any half way sensible formatter / linter is better than no formatter. The rules themselves matter very little, the important part is that rules exist, they are transparent an enforcible and contributors can get it right the first time and there is no nitpicking over whitespace in pull requests.

I agree with @emilazy , just give someone authority over this and get it done. It's not a topic that will be solved in open public conversation. For reasons that are beyond me, people have extremely strong opinions on the topic that can rarely be unified to a degree that really satisfies everyone.